src/Common/CurrentMemoryTracker.cpp

```diff
@@ -51,15 +51,19 @@ void CurrentMemoryTracker::allocImpl(Int64 size, bool throw_if_memory_exceeded)
     {
         if (current_thread)
         {
-            current_thread->untracked_memory += size;
+            Int64 will_be = current_thread->untracked_memory + size;
 
-            if (current_thread->untracked_memory > current_thread->untracked_memory_limit)
+            if (will_be > current_thread->untracked_memory_limit)
             {
                 /// Zero untracked before track. If tracker throws out-of-limit we would be able to alloc up to untracked_memory_limit bytes
                 /// more. It could be useful to enlarge Exception message in rethrow logic.
-                Int64 tmp = current_thread->untracked_memory;
                 current_thread->untracked_memory = 0;
-                memory_tracker->allocImpl(tmp, throw_if_memory_exceeded);
+                memory_tracker->allocImpl(will_be, throw_if_memory_exceeded);
```

Comment:
Now, I think that zeroing untracked_memory also may lead to incorrect accounting.
Technically we should only revert current allocation.

Reply:
Indeed, good catch!
Actually it can be just moved after `allocImpl` and this should fix the problem.

Reply:
Then the comment above will be wrong :)
Tried to increase the limit instead of that.

Reply:
Yep, saw it too.
I'm not the fan of adding more fields into `ThreadStatus` for this, but this looks better then try/catch anyway.